File: leetcode-submissions/1751_Maximum_Number_of_Events_That_Can_Be_Attended_II.cpp
class Solution {
public:
    int n ;
    vector<vector<int>>dp;

    int getIndex(vector<vector<int>>& events,int end, int start ){
        int l = start ;
        int r= n-1;
        int result = n;
        while(l<=r){
            int mid = l+(r-l)/2;
            if(events[mid][0]>end){
                result= mid;
                r=mid-1;
            }else{
                l= mid+1;
            }
        }
        return result ;
    }
    int solve(vector<vector<int>>& events, int k, int i ){
        if(i>=n or k==0) return 0 ;
        if(dp[i][k]!=-1) return dp[i][k];

        int take = events[i][2]+solve(events, k-1, getIndex(events,events[i][1],i));
        int not_take = solve(events , k, i+1);
        return dp[i][k]=max(take , not_take);


    }
    int maxValue(vector<vector<int>>& events, int k) {
        n= events.size();
        dp.clear();
        dp.resize(n+1 , vector<int>(k+1, -1));
        sort(events.begin(), events.end());
        return solve(events , k, 0);

    }
};

File: leetcode-submissions/1751_Maximum_Number_of_Events_That_Can_Be_Attended_II.ts
function maxValue(events: number[][], k: number): number {
    events.sort((a, b)=>a[0]-b[0]);
    const dp:number[][]= new Array(events.length+1).fill(0).map(()=>new Array(k+1).fill(-1));
    return solve(events , k , 0, dp );
};
function getIndex(events : number[][], lastPickedEnd : number):number{
    const n:number = events.length;
    let index:number = n , left : number =0 , right : number = events.length-1;
    while(left<=right){
        const Mid: number = Math.floor((left+right)/2);
        if(events[Mid][0]>lastPickedEnd){
            index = Mid ;
            right = Mid-1;
        }else{
            left = Mid+1;
        }
    }
    return index ;
}
function solve(events:number[][] , k:number, i :number,dp:number[][]):number{
    const n: number = events.length;
    if(i>=n || k===0){
        return 0 ;
    }
    if(dp[i][k]!==-1){
        return dp[i][k];
    }

    const take : number = events[i][2]+solve(events , k-1, getIndex(events, events[i][1]), dp);
    const not_take : number = solve(events , k, i+1, dp);
    return dp[i][k] =  Math.max(take  , not_take);

}